Remove case lists and details from transifex

emord opened the pull request:
@mkangia we had fairly similar ideas for this. This would use a default django admin interface to list out the translations to remove. I'll need to do a bit more testing/documentation on this to make sure it can be used for everything we need, but let me know what you think of this

emord committed:
Added model to blacklist transifex translations

emord committed:
Added admin site for translation blacklist

emord committed:
Added ability to remove translations from transifex via TranslationBlacklist

emord commented on corehq/apps/translations/models.py:
```diff
@@ -92,3 +100,21 @@ def get_translations(cls, lang, key=None, one=False):
+        )
+    )
+    field_name = models.CharField(max_length=255)
+    display_text = models.CharField(max_length=255)
```
this may not be needed, but if it is, it should probably change to a TextField

mkangia replied:
any reason for using `field_name` instead of `case_property`?
`display_text` is not neeeded

emord replied:
display_text was needed because you can have multiple properties displayed in different ways. For example due list displays `parent/parent/dob` as date of birth and the age in years. Ideally we'd use a different property in the module to filter those out, but at this time display_text is all that's passed to this.

mkangia commented on corehq/apps/app_manager/app_translations/generators.py:
```diff
@@ -185,6 +186,29 @@ def _filter_invalid_rows_for_form(self, rows, form_id, label_index):
+    def _blacklisted_translations(self):
+        return TransifexBlacklist.objects.filter(domain=self.domain, app_id=self.app_id).all()
+
+    def _filter_invalid_rows_for_module(self, rows, module_id, case_property_index, list_or_detail_index):
```
Do you plan to refine this logic? 
There are quite some iterations here.

I was thinking of
1. fetch the blacklisted translations by module_id instead, which thus result in more queries
2. when you get the blacklisted translations for the app, iterate over them to create a dict of blacklisted translations per module.

then later you could just do
```
blacklisted_case_properties = self. _blacklisted_translations[module_id]
for i, row in enumerate(rows):
   if row[case_property_index] not in blacklisted_case_properties:
       valid_rows.append(row)
```

mkangia reviewed (approved):
looks great. left one possible improvement in code.
Nice thought on using django admin. Do the GS team have access to it?

emord committed:
Merge branch 'master' into je/case-list-detail-transiex

emord committed:
Add filtering based on display_text

emord committed:
Change to using TextField

emord committed:
Document the transifex blacklist model

emord commented:
@mkangia I did some more testing and added some documentation to the model.

I believe that access to the admin panel is restricted to is_staff at the moment which is just devs. For now I can add their current case list and details to be removed, and I will consider an access plan later

mkangia commented on corehq/apps/translations/generators.py:
```diff
@@ -230,7 +233,9 @@ def occurrence(_row):
                 case_property_index = self._get_header_index(sheet_name, 'case_property')
                 list_or_detail_index = self._get_header_index(sheet_name, 'list_or_detail')
-                rows = self._filter_invalid_rows_for_module(rows, type_and_id.id, case_property_index, list_or_detail_index)
+                default_en_index = self._get_header_index(sheet_name, 'default_en')
```
en would not be the default language always so should do https://github.com/dimagi/commcare-hq/blob/9228f6a6e0e707c1c28cef4f2d98885ef860c001/corehq/apps/translations/generators.py#L197

mkangia commented on corehq/apps/translations/models.py:
```diff
@@ -113,8 +135,9 @@ class TransifexBlacklist(models.Model):
-    display_text = models.CharField(max_length=255)
+    field_name = models.TextField(help_text=FIELD_NAME_HELP)
+    display_text = models.TextField(
+        help_text="The default English translation for this detail/list")
```
this is a bit risky. In case the source lang is changed in the app, it would be skipped. 
One way to avoid that would be to ensure that all blacklisted translations are present at least once? Can't think of a better way for this but definitely worth considering.

Also should we make this optional so that one needs to enter it only why needed to differentiate between similar case list/details?

emord replied:
>  In case the source lang is changed in the app,

That won't be true for ICDS though right? I'm imagining that ICDS is the only one that would be using this  while REACH is being developed

mkangia replied:
we did change the source language from hindi to english on icds. But a couple of projects are looking at using transifex integration and hence might use this as well.

emord replied:
Thanks I didn't realize that, or that other projects would be using transifex (since this used to live in the custom/icds folder). I also didn't see that we use default_lang_index a couple of lines above this, so thanks for pointing that out as well

emord committed:
Make display_text optional

emord committed:
Use the default language instead of English

mkangia commented on corehq/apps/translations/generators.py:
```diff
@@ -186,6 +187,31 @@ def _filter_invalid_rows_for_form(self, rows, form_id, label_index):
+                for trans in self._blacklisted_translations
+                if (
+                    trans.module_id == module_id
+                    and trans.field_type == list_or_detail
```
nit: can rename this to be
`backlisted_trans.field_type == trans_field_type`?

mkangia commented on corehq/apps/translations/models.py:
```diff
@@ -92,3 +100,45 @@ def get_translations(cls, lang, key=None, one=False):
+Example: case detail for tasks_type would have entries:
+    tasks_type (ID Mapping Text)
+    child (ID Mapping Value)
+    pregnancy (ID Mapping Value)
```
nit: can we just say that this should match what the user see under `property` column in bulk translations download so that they can just pick it from there instead of trying to figure out what to put ?

emord replied:
It does seem useful to have that as an example. I added a note about it being the same as in bulk translations above

mkangia commented on corehq/apps/translations/models.py:
```diff
@@ -92,3 +100,45 @@ def get_translations(cls, lang, key=None, one=False):
+    """Used for removing case list and case detail translations before an
+    upload to Transifex.
+
+    This assumes that a default source translation is English exists.
```
this needs to change now?

mkangia commented on corehq/apps/translations/models.py:
```diff
@@ -92,3 +100,45 @@ def get_translations(cls, lang, key=None, one=False):
+    field_name = models.TextField(help_text=FIELD_NAME_HELP)
+    display_text = models.TextField(
+        help_text="The default language's translation for this detail/list. "
+        "If display_text is not filled out then all translations will be blacklisted")
```
`all` can be misinterpreted here as all translations. You can say all translations matching field_type and field_name instead

mkangia reviewed (approved):
Non blocking changes.
Thanks for taking care of this @emord

emord committed:
Update transifex translation documentation

emord commented:
@mkangia thanks for the close review of documentation. I addressed all your feedback in https://github.com/dimagi/commcare-hq/pull/22580/commits/b27985d6898083b95e011197c001e15f79ed17f3

mkangia reviewed (approved):
looks good.

emord committed:
Merge branch 'master' into je/case-list-detail-transiex

emord committed:
Add migration that had conflict from merge with master
